app/src/main/java/pm/gnosis/heimdall/data/repositories/impls/WalletConnectBridgeRepository.kt

```diff
@@ -0,0 +1,463 @@
+            val meta = it.peerMeta()
+            BridgeRepository.SessionEvent.MetaUpdate(
+                BridgeRepository.SessionMeta(
+                    sessionId,
+                    meta?.name,
+                    meta?.description,
+                    meta?.icons,
+                    sessions.containsKey(sessionId),
+                    it.approvedAccounts()?.map { acc -> acc.asEthereumAddress()!! }
+                )
+            )
+        }
+
+    override fun initSession(sessionId: String): Completable = Completable.fromAction {
+        val session = sessions[sessionId] ?: throw IllegalArgumentException("Session not found")
+        session.init()
+    }
+        .subscribeOn(Schedulers.io())
+
+    override fun approveSession(sessionId: String): Completable =
+        safeRepository.observeSafes().firstOrError()
+            .flatMapCompletable { safes ->
+                if (safes.isEmpty()) throw IllegalStateException("No Safe to whitelist")
+                Completable.fromAction {
+                    val session = sessions[sessionId] ?: throw IllegalArgumentException("Session not found")
+                    session.approve(safes.map { it.address.asEthereumAddressChecksumString() }, BuildConfig.BLOCKCHAIN_CHAIN_ID)
+                }
+            }
+            .subscribeOn(Schedulers.io())
+
+    override fun rejectSession(sessionId: String): Completable = Completable.fromAction {
+        val session = sessions[sessionId] ?: throw IllegalArgumentException("Session not found")
+        session.reject()
+    }
+        .subscribeOn(Schedulers.io())
+
+    override fun closeSession(sessionId: String): Completable = Completable.fromAction {
+        val session = sessions[sessionId] ?: throw IllegalArgumentException("Session not found")
+        session.kill()
+    }
+        .subscribeOn(Schedulers.io())
+
+    override fun approveRequest(requestId: Long, response: Any): Completable = Completable.fromAction {
+        val session = sessionForRequest(requestId) ?: throw IllegalArgumentException("Session not found")
+        session.approveRequest(requestId, response)
+    }
+        .subscribeOn(Schedulers.io())
+
+    override fun rejectRequest(requestId: Long, errorCode: Long, errorMsg: String): Completable = Completable.fromAction {
+        val session = sessionForRequest(requestId) ?: throw IllegalArgumentException("Session not found")
+        session.rejectRequest(requestId, errorCode, errorMsg)
+    }
+        .subscribeOn(Schedulers.io())
+
+    override fun activateSession(sessionId: String): Completable = Completable.fromAction {
+        val session = sessionStore.load(sessionId) ?: throw IllegalArgumentException("Session not found")
+        internalCreateSession(session.config)
+        startBridgeService()
+    }
+        .subscribeOn(Schedulers.io())
+
+    private fun startBridgeService() {
+        context.startService(Intent(context, BridgeService::class.java))
+    }
+}
+
+
+interface Session {
```

Comment:
This and other walletconnect logic releated classes should be moved to a separate sdk